File: 0071-simplify-path/0071-simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();
        String[] parts = path.split("/");
        for (String part : parts) {
            if (part.equals("") || part.equals(".")) continue;
            else if (part.equals("..")) {
                if (stack.isEmpty()) continue;
                else stack.pop();
            } else stack.push(part);
        }

        if (stack.isEmpty()) return "/";

        List<String> list = new ArrayList<>(stack);
        Collections.reverse(list);
        return "/" + String.join("/", list);
    }
}

File: 0071-simplify-path/0071-simplify-path.kt
class Solution {
    fun simplifyPath(path: String): String {
        val stack = ArrayDeque<String>()
        val parts = path.split("/")

        for (part in parts) {
            when {
                part == "" || part == "." -> continue
                part == ".." -> {
                    if (stack.isNotEmpty()) stack.removeLast()
                }
                else -> stack.addLast(part)
            }
        }

        if (stack.isEmpty()) return "/"

        return "/" + stack.joinToString("/")
    }
}